packages/tests/server/react/useContext.test.tsx

```diff
@@ -296,6 +297,88 @@ test('invalidate', async () => {
   expect(stableProxySpy).toHaveBeenCalledTimes(2);
 });
 
+test('invalidate procedure for both query and infinite', async () => {
+  const { proxy, App } = ctx;
+  const invalidateQuerySpy = jest.fn();
+  const invalidateInfiniteSpy = jest.fn();
+
+  function MyComponent() {
+    const allPostsList = proxy.post.list.useQuery(
+      { cursor: undefined },
+      {
+        onSuccess: invalidateQuerySpy,
+      },
+    );
+    const allPostsListInfinite = proxy.post.list.useInfiniteQuery(
+      { cursor: undefined },
+      { onSuccess: invalidateInfiniteSpy },
+    );
+
+    const utils = proxy.useContext();
+
+    return (
+      <>
+        <button
+          data-testid="invalidate-button"
+          onClick={() => {
+            utils.post.list.invalidate();
+          }}
+        />
+        <div data-testid="list-status">
+          {allPostsList.isFetching || allPostsListInfinite.isFetching
+            ? 'fetching'
+            : 'done'}
+        </div>
+        <div>
+          {allPostsListInfinite.data?.pages.map((page) => {
+            return page.map((post) => {
+              return <div key={post.id}>{post.text}</div>;
+            });
+          })}
+        </div>
+      </>
+    );
+  }
+
+  const utils = render(
+    <App>
+      <MyComponent />
+    </App>,
+  );
+
+  await waitFor(() => {
+    expect(utils.container).toHaveTextContent('done');
+  });
+
+  await waitFor(() => {
+    expect(utils.container).toHaveTextContent('new post');
+  });
+
+  // I've tried everything to avoid this but for some reason, even though
+  // infinite query is successful and data is on the component it still shows
+  // no calls to the proxy passed to "onSuccess" unless i do this wait ... even
+  // await new Promise(process.nextTick); doesn't work 🤷🤯
+  // https://github.com/trpc/trpc/pull/3195#discussion_r1026952381
+  await waitMs(1);
```

Comment:
We could just use `waitFor`, this LGTM

Reply:
Problem was what do you wait for? 

I was already waiting for the state of the `query` and `infinite` query to not be fetching but at this point the jest proxy function attached to the `onSuccess` of the `infinite` query was still only called 0 (first time) or 1 (after invalidation) times 😵‍💫... However waiting for `1ms` made them read what they should be! (1 then 2) 🤷

As above conversation with TkDodo it seems there is some odd batching going on and the `onSuccess`callback is not qued when fetching is set to false!

Reply:
Ohhh could you `waitFor` the actual proxy count? 🤦

Reply:
That worked and is much neater ... Thanks for pointing out the obvious answer I missed 🤦... Still odd behaviour!...